Support Linking Custom Project Locations

RyanThomas73 opened the pull request:
**Issue #220
* Added `GetReactNativeProjectConfig` helper method/module.
* Updated the `appcenter-link-scripts` for android/ios to use the react-native project config's `android.folder`, `ios.folder`, and `ios.sourceDir` properties to determine the correct project paths when linking.
* Updated the `appcenter-link-sripts` with a new `PbxProject` module and an exported `updateFrameworkSearchPaths(..)` function for ios.
* Updated the `appcenter`, `appcenter-analytics`, and `appcenter-crashes` `prelink.js` scripts to log a message when the android linking logic is being skipped.
* Updated the `appcenter`, `appcenter-analytics`, and `appcenter-crashes` `postlink.js` scripts to log a message when the ios linking logic is being skipped.
* Updated the `appcenter`, `appcenter-analytics`, and `appcenter-crashes` `postlink.js` scripts to update the parent `project.pbxproj` `FRAMEWORK_SEARCH_PATHS` to include the 'Pods/**' frameworks if not already present in the search paths.

RyanThomas73 commented:
Note: I'm still performing some testing on these fixes with some of my projects. I wanted to get the PR submitted early so the maintainers can provide any feedback as soon as possible. I'll update or respond to this comment once I'm done with my additional testing.

guperrot commented:
Here is the reason why CI check fails:

```
2018-01-16T19:51:56.5197180Z > appcenter@1.1.0 lint /Users/vsts/agent/2.126.0/work/1/s/appcenter
2018-01-16T19:51:56.5230670Z > eslint .
2018-01-16T19:51:56.5237560Z 
2018-01-16T19:51:58.5010420Z 
2018-01-16T19:51:58.5028050Z /Users/vsts/agent/2.126.0/work/1/s/appcenter/scripts/postlink.js
2018-01-16T19:51:58.5043180Z    2:7   error  'path' is assigned a value but never used                                                              no-unused-vars
2018-01-16T19:51:58.5058510Z    5:17  error  Unexpected block statement surrounding arrow body; move the returned value immediately after the `=>`  arrow-body-style
2018-01-16T19:51:58.5077080Z   32:21  error  Unexpected block statement surrounding arrow body; move the returned value immediately after the `=>`  arrow-body-style
```

dhei commented:
Hello @RyanThomas73,

Thank you, we really appreciate your PR. Sorry for my delay of response. We have a task in our backlog to review your PR, let me get back to you next week to update the progress.

bmourat commented on appcenter-link-scripts/src/GetReactNativeProjectConfig.js:
```diff
@@ -0,0 +1,9 @@
+
+
+const getReactNativeCliConfig = require('../../../react-native/local-cli/core/index');
```
Shouldn't this be 
```
const getReactNativeCliConfig = require('../../react-native/local-cli/core/index');
```

bmourat commented on appcenter-link-scripts/src/GetReactNativeProjectConfig.js:
```diff
@@ -0,0 +1,9 @@
+const getReactNativeCliConfig = require('../../../react-native/local-cli/core/index');
+
+const GetReactNativeProjectConfig = function () {
+    return getReactNativeCliConfig().getProjectConfig();
```
and this 
```
return getReactNativeCliConfig.getProjectConfig();
```

RyanThomas73 commented:
@bmourat I corrected the issues you pointed out in GetReactNativeProjectConfig.js and a line to require and run react-native's `babelSetup.js` which should address the `SyntaxError: Unexpected token import` error you mentioned.

bmourat commented:
Thanks, I'll test it later!

bmourat commented on appcenter-link-scripts/src/GetReactNativeProjectConfig.js:
```diff
@@ -0,0 +1,11 @@
+const reactNativeCliConfig = require('../../react-native/local-cli/core/index');
+
+const GetReactNativeProjectConfig = function () {
+    return reactNativeCliConfig.getProjectConfig();
```
This line always returns an empty object.

RyanThomas73 replied:
Hi Murat.

To give you more background information on what happens:
The react native cli uses a config object with functions for different configuration values to control much of the cli behavior. One of those functions `getProjectConfig()` contains paths to where the projects are located.

The config object uses a default from within the react-native package. It then loads a config using the below logic and allows any of the functions defined in the loaded config to override the default. The loaded config is found using the following logic:

1) If a --config path/to/config.js is provided it loads the config from the provided path.
e.g. `react-native link --config path/to/custom/react-native-cli-config.js`
**NOTE:** This option does not currently work for the prelink / postlink scripts, as the link command does not pass the config parameter through when calling these scripts.

2) If no --config option is specified it traverses up the directory tree for the nearest ancestor directory looking for a file named `rn-cli.config.js`. If it finds one then it loads that file allowing it to override config logic.

Overriding the values in a `rn-cli.config.js` in the project root or in a config file specified via the --config flag will let the user run the react native cli commands, including the link command, when they have a file structure other than the default.

RyanThomas73 commented:
@bmourat I was unable to reproduce the problem you mentioned where the `getProjectConfig()` returns an empty object. I did, however, find some additional bugs which I have fixed.

I added a test app named `TestAppNonStandardFolderStructure/` with the project folders nested one level. e.g. `TestAppNonStandardFolderStructure/NestedProjectFolder/android/`. The project has an `rn-cli.config.js` file specifying the custom project folder locations.

If you're still able to reproduce the 'empty object' problem you're seeing, could you provide me the steps your using to setup the new project and reproduce?

bmourat commented on appcenter-link-scripts/src/ios/PodFile.js:
```diff
@@ -70,23 +69,17 @@ Podfile.prototype.install = function () {
-        debug(`No podfile found in ${cwd}`);
-        childProcess.execSync('pod init', { cwd });
+Podfile.initializePodfileIfNecessary = function (podfilePath) {
+    if (!fs.statSync(podfilePath).isFile()) {
```
statSync throws an exception when file is not found, thus preventing from creating podfile

RyanThomas73 commented:
Yes, the `react-native run-android` command requires the `--root NestedProjectFolder` flag be provided to work with the non standard folder structure.

The `react-native run-ios` currently requires the `--project-directory NestedProjectFolder/ios/` flag be provided to work with the custom folder structure. Even with this flag the appcenter react-native sdk will most likely fail to build due to Issue #137. Still, having the `react-native link` command working will be a big help productivity wise. I think those of us with non standard structures will just have to continue working around #137 until we can come up with a good solution for that one. 

I'm working on fixing that podfile initialization code. I'll try to test react-native v0.34 to make sure this still works for that version. Do you have any additional suggestions? I'd like to expand the unit tests for the code but I'm not really sure where to start / what tests to begin adding.

RyanThomas73 commented:
@bmourat My updates from yesterday should handle the following:
* Fix the `PodFile.js` error you pointed out.
* Update the `package.json` script aliases in the `TestAppNonStandardFolderStructure/` for the `run-android` and `run-ios` command. The script aliases now included the necessary cli flags for these commands to work.
* Added a workaround for Issue #137 to the `TestAppNonStandardFolderStructure/`.
* Added a fix so the appcenter-link-scripts will work with `react-native v0.34`.

Let me know what you need next from my end.

bmourat commented:
Thanks, I'll test it later!

elamalani commented:
@RyanThomas73 Please address some comments from @bmourat in order for us to review your latest PR. Thanks.

guperrot commented:
Hi, this PR is idle and we will close it now.
Feel free to reopen if you have any updates to address the feedback comments and resolve the conflicts.
